app/models/profile.rb

```diff
@@ -70,5 +70,24 @@ def from_openscap_parser(op_profile, benchmark_id: nil, account_id: nil)
 
       profile
     end
+
+    def rules_and_rule_groups(profile)
+      profile_rules = profile.rules
+      profile_rule_groups = profile.rule_groups
+      custom_arrange(profile_rule_groups, profile_rules)
+    end
+
+    def custom_arrange(rule_groups, rules)
+      all_rule_children = rules.left_outer_joins(:rule_groups).where(rule_groups: { id: rule_groups.select(:id) })
+
+      arranged_rule_groups = rule_groups.arrange_serializable do |parent, children|
+        rule_children = parent.rules.where(rules: { id: rules.select(:id) }).as_json(root: :rule)
+        { rule_group: parent, group_children: children, rule_children: rule_children }
+      end
+
+      parent_rules = rules - all_rule_children
+
+      (arranged_rule_groups + parent_rules.as_json(root: :rule)).to_json
+    end
```

Comment:
@skateman @romanblanco I was working on finding a solution that used Postgres to get rid of most of the arrays mapping/iterations but while I was doing that I came up with this. Does this look better? 

Btw, the `as_json` is only necessary to convert the activerecord result to a hash using the root `:rule` so that it follows the same convention as the `rule_group` hashes

Reply:
Ideally you should have a single JSON conversion at the end :thinking:

Reply:
`as_json` converts ActiveRecord objects to Ruby hashes so it's just getting the rule objects in the right format so that `to_json` will work correctly. Another way of doing it is something like this: 
```
parent_rules.to_a.map() {|x| 
  { :rule => x.serializable_hash() } 
}
```
But this does the same thing as `as_json(root: :rule)` . Does this make sense?

Reply:
Any premature JSON conversion just does a lot of allocations, we should try to minimize those.